### ISSUE TITLE
Double buffering possible?

### ISSUE THREAD

**Issue by cryham** [KurtE/ILI9341_t3n#31]
Hi KurtE
Firstly awesome library!

I am using Teensy 4.0 and ILI9341. Since 320x240 x 2bytes is 150 kB of 512 kB, I could be able to do double buffering. I mean: drawing in one buffer, when done set DMA transfer to update LCD from it, then switch to drawing in second buffer, and so on.
I tried this in few ways but it is not working. Is it possible, am I doing something wrong?
Here is my code:
```c++
uint16_t data[2][320 * 240];  // screen buffers
int buf = 0;

tft.setFrameBuffer(data[0]);
tft.useFrameBuffer(true);

tft.begin();
...

while(1)
{
	tft.setFrameBuffer(data[buf]);
	buf = 1-buf;

	... drawing stuff ....

	tft.waitUpdateAsyncComplete();
	tft.updateScreenAsync();
}
```
Full code [here](https://github.com/cryham/kc4/blob/master/kc/main.cpp) single buffer, without that above.

**Comment by KurtE** [comment]
Sorry, not sure if this was sort of already answered on the forum or not? 

Sometimes my github email gets eaten by my email providers spam system. 

I just pushed up a change that maybe helps... That is the code to update the DMA buffers was not checking that I had cleared the DMA had been initialized, only that the whole field was not zero. 

Let me know if this helps.  Again I more likely to see things when posted up on forum.

Again Sorry I did not respond until now.

**Comment by cryham** [comment]
Thanks.
I tested but your changes didn't fix it. This time I have black screen all the time and it crashes.
Previously I saw few frames flickering with black screen and it crashed.
It's bad for testing since I have to hold reset for 10sec or more after it crashes to get Teensy back.
BTW. I am not using RST for tft (it has Vcc on pin) not sure if matters.
I didn't know of the forum topic. Right, there is a way to do double buffering with memcpy, likely slower  [on page 13](https://forum.pjrc.com/threads/38752-Playing-around-with-yet-another-version-of-ILI9341_t3-library-(ILI9341_t3n)/page13).

**Comment by KurtE** [comment]
I just pushed up a simple test sketch and code that switches back and forth between two buffers.

Let me know if this fixes the problem.

**Comment by cryham** [comment]
I tested your new doubleBuffer example in Arduino IDE on my setup.
This is the output from Serial Monitor:
```
ILI9341 Test!
Delta time: 42885
Press any key to continue
Continued
Delta time: 42884
Press any key to continue
Continued
Delta time: 42974
Press any key to continue
Continued
Delta time: 42971
Press any key to continue
Continued
.
............
```
It did show screens with "Using FB1", I pressed enter then FB2, FB1A and FB2A.
Then it started drawing green lines but stopped after about 12 (dots). So the display has top few lines green, rest is black and nothing happens.
So I think I have some problem. Is this a hardware issue or software? This is exactly what happened before in my code it also just stopped after few frames swtiched. It looks for me like it was waiting infinitely in function updateScreenAsync(), but it doesn't show Timeout before.
Any ideas?
